### PR TITLE
Add CHIP-SPV integration

### PR DESCRIPTION
**Pull request by pvelesko**
First touch for integrating SPIR-V based backend CHIP-SPV. Still a WIP but so far it's worked just as well as the Perl-based hipcc. 

Also, there are some changes to shared and AMD-specific components along with some comments that I would like you to review.

### CODE REVIEW

**Comment by searlmc1** [comment]
Hi, 

Would you please submit this to 'develop' branch?

Thanks

**Comment by yxsamliu** [comment]
We plan to moving all the logics from hipcc to clang and finally deprecate hipcc or make it a no-op wrapper.

I am not sure it is a good idea to add stuff to hipcc.

It is better to use clang directly to compile for CHIP-SPV.

**Comment by pvelesko** [comment]
> We plan to moving all the logics from hipcc to clang and finally deprecate hipcc or make it a no-op wrapper.


on the README for this very repository it says that the plan is to deprecate the Perl based hipcc and move to using this instead. Are you saying that's no longer the plan? @yxsamliu

File: src/hipBin.cpp [hunk 1]
```diff
@@ -59,20 +62,32 @@ HipBin::HipBin() {
...
   if (hipBinAMDPtr_->detectPlatform()) {
     // populates the struct with AMD info
     const PlatformInfo& platformInfo = hipBinAMDPtr_->getPlatformInfo();
     platformVec_.push_back(platformInfo);
     hipBinBasePtrs_.push_back(hipBinAMDPtr_);
     platformDetected = true;
-  } else if (hipBinNVPtr_->detectPlatform()) {
+  } 
+  
+  if (hipBinNVPtr_->detectPlatform()) {
```
Comment (yxsamliu) [inline src/hipBin.cpp:75]: if -> else if

File: src/hipBin.cpp [hunk 2]
```diff
@@ -59,20 +62,32 @@ HipBin::HipBin() {
...
     // populates the struct with Nvidia info
     const PlatformInfo& platformInfo = hipBinNVPtr_->getPlatformInfo();
     platformVec_.push_back(platformInfo);
     hipBinBasePtrs_.push_back(hipBinNVPtr_);
     platformDetected = true;
+  } 
+  
+   if (hipBinSPIRVPtr_->detectPlatform()) {
```
Comment (yxsamliu) [inline src/hipBin.cpp:83]: if -> else if

Comment (pvelesko) [reply src/hipBin.cpp:83]: If we do this then we stop at the first detected platform and do not completely populate the platform vector. 

In `HipBin::executeHipCC` we use the 0th platform for now so nothing should break.
In `HipBin::executeHipConfig` we iterate and print through all the available platforms.

**Comment by scchan** [comment]
> > We plan to moving all the logics from hipcc to clang and finally deprecate hipcc or make it a no-op wrapper.
> 
> on the README for this very repository it says that the plan is to deprecate the Perl based hipcc and move to using this instead. Are you saying that's no longer the plan? @yxsamliu

hipcc was created in the past to improve the user experience because at that time, the HIP language in clang was seriously lacking.  That is no longer the case today.  I'd strongly recommend new users and platforms coming to HIP to consider using clang directly instead of hipcc.  We are moving to the hipcc C++ version mainly to support existing users who already have a dependence on hipcc and to improve the robustness of the tool.

**Comment by pvelesko** [comment]
> We are moving to the hipcc C++ version mainly to support existing users who already have a dependence on hipcc and to improve the robustness of the tool.

As a HIP implementation, we would also need to maintain backwards compatibility for codes that rely on hipcc.  @scchan

**Comment by scchan** [comment]
> > We are moving to the hipcc C++ version mainly to support existing users who already have a dependence on hipcc and to improve the robustness of the tool.
> 
> As a HIP implementation, we would also need to maintain backwards compatibility for codes that rely on hipcc. @scchan

ok, can you split the integration of the SPIRV support into a couple of smaller PRs?
Some other changes seem to be general bug fixes and aren't directly related to SPIRV so could you put those into different PRs please?

**Comment by kzhuravl** [comment]
Do we still need this PR? If yes, can you rebase it?

**Comment by pvelesko** [comment]
You can close for now if needed, I'll reopen

**Comment by kzhuravl** [comment]
Thanks, closing

**Pull request closed without merging**